### PR TITLE
Heapsort

### LINKED ISSUE #107: Adding `sort` method to heaps

### PR DESCRIPTION
**Pull request by HarsheetKakar**
#### References to other Issues or PRs or Relevant literature

Fixes #107 

#### Brief description of what is fixed or changed
Used min heap to make heapsort. 
Passes all tests.

### CODE REVIEW

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -235,6 +236,30 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    end = kwargs.get('end', len(array) - 1)
+
+    h = BinaryHeap(heap_property="min")
+    for i in range(start, end+1):
+        if array[i] is not None:
+            h.insert(array[i], array[i])
+        array[i] = None
+
+    k = start
+    while(True):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:252]:
```suggestion
    while True:
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -235,6 +236,30 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+        try:
+            x = h.extract()
+            array[k] = x.key
+            k += 1
+        except IndexError:
+            break
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:258]: Replace it with `if else`.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
     'merge_sort_parallel',
     'brick_sort',
     'brick_sort_parallel',
+    'heap_sort',
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:11]:
```suggestion
    'heapsort',
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 4]
```diff
@@ -235,6 +236,30 @@ def brick_sort_parallel(array, num_threads, **kwargs):
     if _check_type(array, DynamicArray):
         array._modify(force=True)
 
+def heap_sort(array, **kwargs):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:239]:
```suggestion
def heapsort(array, **kwargs):
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from pydatastructs import (
     merge_sort_parallel, DynamicOneDimensionalArray,
     OneDimensionalArray, brick_sort, brick_sort_parallel,
-    matrix_multiply_parallel)
+    heap_sort, matrix_multiply_parallel)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:4]:
```suggestion
    heapsort, matrix_multiply_parallel)
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 6]
```diff
@@ -50,6 +50,9 @@ def test_brick_sort():
 def test_brick_sort_parallel():
     _test_common_sort(brick_sort_parallel, num_threads=3)
 
+def test_heap_sort():
+    _test_common_sort(heap_sort)
+
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:55]:
```suggestion
def test_heapsort():
    _test_common_sort(heapsort)

```

**Comment by czgdp1807** [comment]
Please add doc strings as well.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 2 above)
Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py:258]: ok

File: pydatastructs/linear_data_structures/algorithms.py (hunk 4 above)
Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py:239]: we have brick sort as brick_sort

**Comment by HarsheetKakar** [comment]
Should we add `reverse` parameter since there is no way to implement `comp`?

File: pydatastructs/linear_data_structures/algorithms.py [hunk 7]
```diff
@@ -235,6 +236,52 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    end = kwargs.get('end', len(array) - 1)
+
+    h = BinaryHeap(heap_property="min")
+    for i in range(start, end+1):
+        if array[i] is not None:
+            h.insert(array[i], array[i])
+        array[i] = None
+
+    i = start
+    while h._last_pos_filled != -1:
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:278]:
```suggestion
    while not h.is_empty:
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 8]
```diff
@@ -235,6 +236,52 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    """
+    from pydatastructs.trees.heaps import BinaryHeap
+
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+
+    h = BinaryHeap(heap_property="min")
+    for i in range(start, end+1):
+        if array[i] is not None:
+            h.insert(array[i], array[i])
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:274]:
```suggestion
            h.insert(array[i])
```

**Comment by czgdp1807** [comment]
> Should we add `reverse` parameter since there is no way to implement `comp`?

 Ignore this thing. The algorithm itself is fixed since it uses a fixed type of heap.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 8 above)
Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py:274]: insert takes 2 arguments, key and data

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py:274]: `def insert(self, key, data):`

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:274]: `data` should be optional for consistency, at all places, some of them are done, may be in a different patch. `key` should be required.

**Comment by HarsheetKakar** [comment]
> > Should we add `reverse` parameter since there is no way to implement `comp`?
> 
> Ignore this thing. The algorithm itself is fixed since it uses a fixed type of heap.

don't you think that comp should be added in heap also, since we already have a class attribute `_comp` inside DHeap

**Comment by czgdp1807** [comment]
> don't you think that comp should be added in heap also, since we already have a class attribute `_comp` inside DHeap

I am not sure if accessing `_comp` from inside heaps will be a good idea as they are only there for combing the code for `min` and `max` heaps. We don't know how it will work for they inputs of `heapsort`.  May be user passes `_comp` with some other kinds of input which are not compatible with heaps then it will create unpredictable problems.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 8 above)
Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py:274]: should I create another PR for that or push it in here?

**Comment by HarsheetKakar** [comment]
> > don't you think that comp should be added in heap also, since we already have a class attribute `_comp` inside DHeap
> 
> I am not sure if accessing `_comp` from inside heaps will be a good idea as they are only there for combing the code for `min` and `max` heaps. We don't know how it will work for they inputs of `heapsort`. May be user passes `_comp` with some other kinds of input which are not compatible with heaps then it will create unpredictable problems.

lets leave it then

**Comment by HarsheetKakar** [comment]
Is it ok to merge?

File: pydatastructs/linear_data_structures/algorithms.py (hunk 8 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:274]: > should I create another PR for that or push it in here?

Create a separate PR and make data optional wherever, `key, data` combination is passed as an argument.

**Comment by HarsheetKakar** [comment]
anything to update in this? @czgdp1807

File: pydatastructs/linear_data_structures/algorithms.py [hunk 9]
```diff
@@ -235,6 +236,52 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+        which is to be sorted.
+        Optional, by default 0
+    end: int
+        The ending index of the portion which
+        is to be sorted.
+        Optional, by default the index
+        of the last position filled.
+
+    Examples
+    ========
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:259]:
```suggestion
    ========

```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 10]
```diff
@@ -235,6 +236,52 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    >>> from pydatastructs import OneDimensionalArray, heapsort
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> heapsort(arr)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:264]:
```suggestion
    [1, 2, 3]
    
    References
    =========
    
    .. [1] https://en.wikipedia.org/wiki/Heapsort
    
    Note
    ====
    
    This function does not support custom comparators as is the case with 
    other sorting functions in this file.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 11]
```diff
@@ -235,6 +236,64 @@ def brick_sort_parallel(array, num_threads, **kwargs):
     if _check_type(array, DynamicArray):
         array._modify(force=True)
 
+def heapsort(array, **kwargs):
+    """
+    Implements Heap Sort sorting algorithm
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:241]:
```suggestion
    Implements Heapsort algorithm.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 12]
```diff
@@ -235,6 +236,64 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    ========
+
+    >>> from pydatastructs import OneDimensionalArray, heapsort
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> heapsort(arr)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
+    
+    References
+    =========
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:268]:
```suggestion
    ==========
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 13]
```diff
@@ -235,6 +236,64 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    
+    .. [1] https://en.wikipedia.org/wiki/Heapsort
+    
+    Note
+    ====
+    
+    This function does not support custom comparators as is the case with
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:275]:
```suggestion
    This function does not support custom comparators as is the case with
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 14]
```diff
@@ -235,6 +236,64 @@ def brick_sort_parallel(array, num_threads, **kwargs):
...
+    References
+    ==========
+    
+    .. [1] https://en.wikipedia.org/wiki/Heapsort
+    
+    Note
+    ====
+    
+    This function does not support custom comparators as is the case with
+    other sorting functions in this file.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:276]:
```suggestion
    [1, 2, 3]

    References
    ==========

    .. [1] https://en.wikipedia.org/wiki/Heapsort

    Note
    ====

    This function does not support custom comparators as is the case with
    other sorting functions in this file.
```

**Pull request merged**